presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java

```diff
@@ -0,0 +1,419 @@
+                        .putAll(subqueryProjection.get().getAssignments())
+                        .build();
+
+                return new ProjectNode(
+                        idAllocator.getNextId(),
+                        aggregationNode.get(),
+                        assignments);
+            }
+            else {
+                return aggregationNode.get();
+            }
+        }
+
+        private Optional<AggregationNode> createAggregationNode(
+                AggregationNode scalarAggregation,
+                JoinNode leftOuterJoin,
+                Symbol nonNullableAggregationSourceSymbol)
+        {
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregations = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Signature> functions = ImmutableMap.builder();
+            FunctionRegistry functionRegistry = metadata.getFunctionRegistry();
+            for (Map.Entry<Symbol, FunctionCall> entry : scalarAggregation.getAggregations().entrySet()) {
+                FunctionCall call = entry.getValue();
+                QualifiedName count = QualifiedName.of("count");
+                Symbol symbol = entry.getKey();
+                if (call.getName().equals(count)) {
+                    aggregations.put(symbol, new FunctionCall(
+                            count,
+                            ImmutableList.of(nonNullableAggregationSourceSymbol.toSymbolReference())));
+                    List<TypeSignature> scalarAggregationSourceTypeSignatures = ImmutableList.of(
+                            symbolAllocator.getTypes().get(nonNullableAggregationSourceSymbol).getTypeSignature());
+                    functions.put(symbol, functionRegistry.resolveFunction(
+                            count,
+                            scalarAggregationSourceTypeSignatures,
+                            false));
+                }
+                else {
+                    aggregations.put(symbol, entry.getValue());
+                    functions.put(symbol, scalarAggregation.getFunctions().get(symbol));
+                }
+            }
+
+            List<Symbol> groupBySymbols = leftOuterJoin.getLeft().getOutputSymbols();
+            return Optional.of(new AggregationNode(
+                    idAllocator.getNextId(),
+                    leftOuterJoin,
+                    groupBySymbols,
+                    aggregations.build(),
+                    functions.build(),
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+
+        public Optional<RewrittenScalarAggregationSource> rewriteScalarAggregationSource(PlanNode aggregationSource, List<Symbol> correlation)
```

Comment:
There's nothing specific to this being an "aggregation source", is it? Looking at the code below, this appears to be doing something that can be described as "decorrelating filters". I would just call it "decorrelateFilters" and change the name of the parameter to `node`.